### DISCUSSION TITLE
com.arcadedb.exception.TimeoutException: Timeout on locking file

### DISCUSSION THREAD

**Discussion by happy-kitty** [ArcadeData/arcadedb#1157] [General]
Hi.
We're trying to perform an initial graph population (i.e. mainly vertex/edge creation) in the fastest possible way minimizing/eliminating concurrent page modifications. There is a single vertex supertype (called _MN_) which has a single unique index. The supertype is essentially "abstract" - we only create vertices in subtypes. There are also 3 "abstract" edge supertypes, but this is probably irrelevant.

We tried to take advantage of the huge number of buckets ArcadeDB claims to support (2B). As each type can have at most 32 buckets, this obviously means a lot of types - which is fine as our objects (modeled as vertices) fall into one of  >25,000 categories.
So our 1st naive attempt was introducing a vertex type for each object category (that would inherit from the vertex super-type) - with a single bucket per type. This didn't work - after creating a few thousand types all schema operations (like adding more types) became intolerably slow. Could it be because all bucket files are stored in a single directory - which doesn't like having tens of thousands of files in it? This brings my 1st question: **is there a way to specify a location for the bucket file in the file system?** 

In our 2nd attempt we decided to have just a few vertex subtypes (3 to start with) and assign each object category to one of these subtypes. Subtypes still inherit from the supertype (and its index). Each subtype is allocated 30 buckets and we make sure that just one thread can write into any bucke. So we create vertices in 3 subtypes in the total of 3*30=90 buckets. This would hopefully give us the desired degree of concurrency. But now we ran into tons of exceptions looking like this:
**com.arcadedb.exception.TimeoutException: Timeout on locking file 4** (_MN__0_1747606273328823.4.64000.v1.umtidx) during commit (fileIds=[4, 12, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, 132, 223, 224, 225, 226, 227, 228, 229, 230, 231, 232, 233, 234, 235, 236, 237, 238, 239, 240, 241, 242, 243, 244, 245, 246, 247, 248, 249, 250, 251, 252, 343, 344, 345, 346, 347, 348, 349, 350, 351, 352, 353, 354, 355, 356, 357, 358, 359, 360, 361, 362, 363, 364, 365, 366, 367, 368, 369, 370, 371, 372, 463, 464, 465, 466, 467, 468, 469, 470, 471, 472, 473, 474, 475, 476, 477, 478, 479, 480, 481, 482, 483, 484, 485, 486, 487, 488, 489, 490, 491, 492, 493, 496, 499, 502, 503, 505, 508, 511, 514, 517, 520, 523, 526, 529, 532, 535, 538, 541, 544, 547, 550, 553, 556, 559, 562, 565, 568, 571, 574, 577, 580, 583, 584, 585, 586, 587, 588, 589, 590, 591, 592, 593, 594, 595, 596, 597, 598, 599, 600, 601, 602, 603, 604, 605, 606, 607, 608, 609, 610, 611, 612, 673, 674, 675, 676, 677, 678, 679, 680, 681, 682, 683, 684, 685, 686, 687, 688, 689, 690, 691, 692, 693, 694, 695, 696, 697, 698, 699, 700, 701, 702, 703, 704, 705, 706, 707, 708, 709, 710, 711, 712, 713, 714, 715, 716, 717, 718, 719, 720, 721, 722, 723, 724, 725, 726, 727, 728, 729, 730, 731, 732]).
From looking at com.arcadedb.database.TransactionIndexContext method addFilesToLock it looks Arcade tries to lock all index files for supertype and all buckets of all of its subtypes - even though we're inserting a vertex into a specific bucket of the specific subtype. 
**What are we doing wrong and how can we solve the problem?**

Thanks in advance.
We tried using Arcade versions 23.3.1, 23.4.1 and 23.6.1.

**Comment by lvca** [comment]
Hello @happy-kitty, are you using a unique index on the base type? That would explain why it needs to lock all the files at commit time index uniqueness).

**Reply by happy-kitty** [reply]
Thanks for the prompt response @lvca 
Yes, the index is unique. But don't subtypes bucket indices contain only the entries in the bucket? Why lock all other buckets?

**Reply by lvca** [reply]
Because the ID is unique across all subtypes, so it must be locked to guarantee it's unique. If you want to go in parallel you should set PARTITION bucket selection, it should reduce the amount of involved files:

```sql
ALTER TYPE Account BucketSelectionStrategy `partitioned('id')`
```

**Reply by happy-kitty** [reply]
OK, great, we'll give this a try

**Reply by happy-kitty** [reply]
And what about my 1st question about thousands of types?

**Reply by lvca** [reply]
The main limitations of having thousands of files are the too many files open + with indexes you have a higher cost + locking all those files is expensive and kills concurrency if you have a unique index.

Are you doing early optimization or did you already try to insert data and it wasn't fast enough with default settings?

**Reply by happy-kitty** [reply]
With default settings the performance was abysmal due to tons of concurrent modifications exceptions

**Reply by happy-kitty** [reply]
I have to be more specific. We originally had no inheritance at all, just a single vertex types with a single unique index and a round-robin bucket selection strategy. This produced endless concurrent page modification exceptions. So we had to really aggressively serialize by object category but this really slowed things down.

**Comment by lvca** [comment]
Try to go back to the simple initial setup, but this time use `thread` bucket selection strategy or `partitioned(id)` to have no concurrency.

**Comment by happy-kitty** [comment]
We effectively have Thread strategy now as we control the buckets for all new documents and assign buckets to the threads. We'll try to play with partitioned

**Reply by lvca** [reply]
If you can send some piece of code or a way to reproduce it ;-)

**Comment by happy-kitty** [comment]
This may be not simple. We don't run into this in simple unit tests but only in massive run of the whole system when a lot of code is involved.
FYI, we're trying to migrate from Neo4J to ArcadeDb.
I'll keep you updated.

**Reply by lvca** [reply]
Ok. Inserting vertices shouldn't be a problem do it in parallel (with the thread or partitioned strategy). With the edge is usually tricky, but it depends on the use case.

In your use case, is the massive insertion more important than the queries later? If this is the case, better to set the number of buckets = to the number of cores, but at the same time it will be slower with queries later. On the opposite, the less buckets you use, the less you can insert in parallel, but lookup from indexes will be faster.

**Reply by happy-kitty** [reply]
It's a good question and unfortunately I don't have a clear answer: both the initial graph population and the queries are important. I actually hoped that we could speed up the queries when we know the bucket (in some cases we do) i.e. searching for a specific bucket of a specific subtype rather than the (polymorphic) supertype search.

**Reply by lvca** [reply]
If you search by an id, and this id property is indexed with `partitioned(id)` strategy, then ArcadeDB knows exactly in which bucket to look for it, because the bucket is calculated from the id (the formula is `bucketId = hash(id) % total_buckets`).

**Reply by happy-kitty** [reply]
Hm, that's what I'm struggling with. What we have right now is effectively a custom bucket selection strategy based on the object category (known at the object creation type but not always at the time of the queries - that's why we have a unique index in the supertype). The value of the indexed property itself (or its hash in the partitioned strategy) is too random and doesn't let us add other objects in the transaction that would share the same bucket without concurrent modifications.
